File: books/cs/introduction-to-algorithms/chapter8/R/counting_sort.R
#
# counting sort algorithm
# introduction to algorithm, 3rd edition, chapter 3
#
# clean the workspace and memory

rm( list=ls() )
gc()

aaa <- sample(10,10,replace = T)

count_sort <- function( aaa ){
  k <- max(aaa)
  bbb <- NULL
  ccc <- rep(0,k)

  for(idx in aaa){
    ccc[idx] <- ccc[idx] + 1
  }
  ccc <- cumsum(ccc)

  aaar <- rev(aaa)
  for(idx in aaar){
    bbb[ccc[idx]] <- idx
    ccc[idx] <- ccc[idx] - 1
  }
  return(bbb)
}

bbb <- count_sort(aaa)

File: books/cs/introduction-to-algorithms/chapter8/Scala/counting_sort.scala
import scala.util.Random

object CountingSort{
        def main(args: Array[String]):Unit = {
                val dataSetSize = 100
                val aaa = List.fill(10)(Random.nextInt(10))
                var bbb = new Array[Int](aaa.size)

                println("the array aaa")
                println(aaa)

                counting_sort(aaa,bbb);
                println(bbb.toList)
        }

        def counting_sort(aaa: List[Int], bbb: Array[Int]):Unit = {
                var ccc = new Array[Int](aaa.max+1)
                var idx = 0
                var total = 0

                for( idx <- aaa ){
                        ccc(idx) = ccc(idx) + 1
                }
                println(ccc.toList)
                total = 0
                for( idx <- 0 to (ccc.size-1) ){
                        total = total + ccc(idx)
                        ccc(idx) = total
                }
                println(ccc.toList)

                for(idx <- aaa.reverse){
                        bbb(ccc(idx)-1) = idx
                        ccc(idx) = ccc(idx) - 1
                }
        }
}
